c++/src/kj/compat/http-test.c++

```diff
@@ -3065,6 +3065,193 @@ KJ_TEST("adapted client/server propagates request exceptions like non-adapted cl
+  doDelayedCompletionTest(false, nullptr);
+}
+
+KJ_TEST("adapted client propagates throw from service after complete response body sent") {
+  doDelayedCompletionTest(true, uint64_t(3));
+}
+
+KJ_TEST("adapted client propagates throw from service after incomplete response body sent") {
+  doDelayedCompletionTest(true, uint64_t(6));
+}
+
+KJ_TEST("adapted client propagates throw from service after chunked response body sent") {
+  doDelayedCompletionTest(true, nullptr);
+}
+
+class DelayedCompletionWebSocketHttpService final: public HttpService {
+public:
+  DelayedCompletionWebSocketHttpService(HttpHeaderTable& table, bool closeUpstreamFirst)
+      : table(table), closeUpstreamFirst(closeUpstreamFirst) {}
+
+  kj::Promise<void> request(
+      HttpMethod method, kj::StringPtr url, const HttpHeaders& headers,
+      kj::AsyncInputStream& requestBody, Response& response) override {
+    KJ_ASSERT(headers.isWebSocket());
+
+    auto ws = response.acceptWebSocket(HttpHeaders(table));
+    kj::Promise<void> promise = kj::READY_NOW;
+    if (closeUpstreamFirst) {
+      // Wait for a close message from the client before starting.
+      promise = promise.then([&ws = *ws]() { return ws.receive(); }).ignoreResult();
+    }
+    promise = promise
+        .then([&ws = *ws]() { return ws.send("foo"_kj); })
+        .then([&ws = *ws]() { return ws.close(1234, "closed"_kj); });
+    if (!closeUpstreamFirst) {
+      // Wait for a close message from the client at the end.
+      promise = promise.then([&ws = *ws]() { return ws.receive(); }).ignoreResult();
+    }
+    return promise.attach(kj::mv(ws)).then([this]() {
+      return kj::mv(paf.promise);
+    });
+  }
+
+  kj::PromiseFulfiller<void>& getFulfiller() { return *paf.fulfiller; }
+
+private:
+  HttpHeaderTable& table;
+  bool closeUpstreamFirst;
+  kj::PromiseFulfillerPair<void> paf = kj::newPromiseAndFulfiller<void>();
+};
+
+void doDelayedCompletionWebSocketTest(bool exception, bool closeUpstreamFirst) noexcept {
+  KJ_HTTP_TEST_SETUP_IO;
+
+  HttpHeaderTable table;
+
+  DelayedCompletionWebSocketHttpService service(table, closeUpstreamFirst);
+  auto client = newHttpClient(service);
+
+  auto resp = client->openWebSocket("/", HttpHeaders(table)).wait(waitScope);
+  auto ws = kj::mv(KJ_ASSERT_NONNULL(resp.webSocketOrBody.tryGet<kj::Own<WebSocket>>()));
+
+  if (closeUpstreamFirst) {
+    // Send "close" immediately.
+    ws->close(1234, "whatever"_kj).wait(waitScope);
+  }
+
+  // Read "foo" from the WebSocket: works
+  {
+    auto msg = ws->receive().wait(waitScope);
+    KJ_ASSERT(msg.is<kj::String>());
+    KJ_ASSERT(msg.get<kj::String>() == "foo");
+  }
+
+  kj::Promise<void> promise = nullptr;
+  if (closeUpstreamFirst) {
+    // Receiving the close hangs.
+    promise = ws->receive()
+        .then([](WebSocket::Message&& msg) { KJ_EXPECT(msg.is<WebSocket::Close>()); });
+    KJ_EXPECT(!promise.poll(waitScope));
```

Comment:
This line should be outside of the if-else block to test the other branch correctly, I think.

Reply:
Whoa, nice catch... When I read test code my eyes just glaze over.

Reply:
It's my favorite pastime. :)